orion/primitives/tadgan.py

```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
-        length = shape[0]
+        length, input_dim = shape
         target_shape = np.asarray(y)[0].shape
+        output_dim = target_shape[1]
 
-        # to infer the shape
+        # Infers the shape.
         self.shape = self.shape or shape
         self.target_shape = self.target_shape or target_shape
-
-        self._setdefault(kwargs, 'generator_reshape_dim', length // 2)
-        self._setdefault(kwargs, 'generator_reshape_shape', (length // 2, 1))
-        self._setdefault(kwargs, 'encoder_reshape_shape', self.latent_shape)
-
+        self.latent_shape = (self.latent_dim, output_dim)
+
+        kwargs.update({
+            'shape': self.shape,
+            'target_shape': self.target_shape,
+            'input_dim': input_dim,
+            'output_dim': output_dim,
+            'generator_reshape_dim': length // 2,
+            'generator_reshape_shape': (length // 2, output_dim),
+            'encoder_reshape_shape': self.latent_shape
+        })
         return kwargs
 
-    def _set_shapes(self):
+    def _set_shapes(self) -> None:
         self.encoder_input_shape = self.shape
         self.generator_input_shape = self.latent_shape
         self.critic_x_input_shape = self.target_shape
         self.critic_z_input_shape = self.latent_shape
 
-    def _build_tadgan(self, **kwargs):
+    def _format_losses(self, losses: list) -> dict:
+        """Format losses into dictionary mapping metric names to their current value."""
+        output = dict()
+        if self.detailed_losses:
+            for i in range(len(losses)):
+                for j in range(len(losses[i])):
+                    output[LOSS_NAMES[i][j]] = losses[i][j]
+
+        else:
+            for i in range(len(losses)):
+                output[LOSS_NAMES[i][0]] = losses[i][0]
+
+        return output
+
+    @staticmethod
+    def _wasserstein_loss(y_true, y_pred):
+        return tf.reduce_mean(y_true * y_pred)
+
+    @staticmethod
+    def _gradient_penalty_loss_wrapper(critic):
+        def _gradient_penalty_loss(real, fake):
+            # Random weighted average to create interpolated signals.
+            batch_size = tf.shape(real)[0]
+            alpha = tf.random.uniform([batch_size, 1, 1], dtype=tf.float64)
+            interpolated = (alpha * real) + ((1 - alpha) * fake)
+
+            # Get the critic output for this interpolated signal.
+            with tf.GradientTape() as gp_tape:
+                gp_tape.watch(interpolated)
+                validity_interpolated = critic(interpolated)
+
+            # Calculate the gradients w.r.t to this interpolated signal.
+            grads = gp_tape.gradient(validity_interpolated, [interpolated])[0]
+
+            # Calculate the norm of the gradients.
+            norm = tf.sqrt(tf.reduce_sum(tf.square(grads), axis=np.arange(1, len(grads.shape))))
+            gradient_penalty = tf.reduce_mean((norm - 1.0) ** 2)
+            return gradient_penalty
+
+        return _gradient_penalty_loss
```

Comment:
minor: for static methods, can we place them directly before `__init__` function? I think that's the proper convention of where to place it in a class

Reply:
Moved stated methods to before `__init__`